rio/src/RIO/Prelude/RIO.hs

```diff
@@ -35,3 +48,60 @@ instance MonadUnliftIO (RIO env) where
 instance PrimMonad (RIO env) where
     type PrimState (RIO env) = PrimState IO
     primitive = RIO . ReaderT . const . primitive
+
+data SomeRef a
+  = SomeRef !(IO a) !(a -> IO ())
+
+readSomeRef :: SomeRef a -> IO a
+readSomeRef (SomeRef x _) = x
+
+writeSomeRef :: SomeRef a -> a -> IO ()
+writeSomeRef (SomeRef _ x) = x
+
+modifySomeRef :: SomeRef a -> (a -> a) -> IO ()
+modifySomeRef (SomeRef read write) f =
+  (f <$> read) >>= write
+
+ioRefToSomeRef :: IORef a -> SomeRef a
+ioRefToSomeRef ref = do
+  SomeRef (readIORef ref)
+          (\val -> atomicModifyIORef' ref (\old -> (val, ())))
```

Comment:
I don't believe the `atomic` adds anything here, since you're ignoring the old value anyway.